### PR TITLE
Biometric 2FA in a Django Application with TypingDNA

### PR DESCRIPTION
**Pull request by Nomzy-kush**
### Article checklist
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by paulodhiambo** [comment]
@Nomzy-kush amazing article. Few things to correct before I proceed with the review.
1. Renaming the images, give them descriptive names.
2. Adding hero image.
3. Rename the image with your author profile image to avatar.
4. Your author profile description should be in `index.md` file not `index.md.TXT` file.

**Review by paulodhiambo — COMMENTED** [review]
@Nomzy-kush amazing article. Few things to correct before I proceed with the review.
1. Renaming the images, give them descriptive names.
2. Adding hero image.
3. Rename the image with your author profile image to avatar.
4. Your author profile description should be in `index.md` file not `index.md.TXT` file.

**Comment by Nomzy-kush** [comment]
@paulodhiambo I have made the required changes please review.

File: articles/biometric-2fa-in-a-django-application-with-typingdna/index.md [hunk 1]
```diff
@@ -0,0 +1,1165 @@
...
+    page_number = request.GET.get('page')
+
+    page_obj = paginator.get_page(page_number)
+
+    context={"page_obj":page_obj}
+
+    return render(request,"videos.html",context)
+
+```
+Also we need to update the `register` to create new `UserProfile` objects for every registered user. To do this, update the `register` view by entering code below after the line with`user.save`:
```
Comment (paulodhiambo) [inline articles/biometric-2fa-in-a-django-application-with-typingdna/index.md:216]: Consider changing "Also we need to update..." to "Also, we need to update..."

File: articles/biometric-2fa-in-a-django-application-with-typingdna/index.md [hunk 2]
```diff
@@ -0,0 +1,1165 @@
...
+        return hashlib.sha1((text + text[::-1]).encode()).hexdigest()
+
+
+```
+
+TypingDNA requires a user ID to save typing patterns to a user, so we created a `hash_text` method to hash the user’s ID before sending.
+
+## Saving Users Typing Patterns
+
+After building our helper library, we can start saving our users’ typing patterns in our accounts for future matching. Lets first import the helper library class into our Django application and create an instance of our helper class by supplying the `apiKey` and `apiSecret` we saved from our TypingDNA dashboard earlier.
```
Comment (paulodhiambo) [inline articles/biometric-2fa-in-a-django-application-with-typingdna/index.md:577]: Change "Lets first import" to "Let's first import"

File: articles/biometric-2fa-in-a-django-application-with-typingdna/index.md [hunk 3]
```diff
@@ -0,0 +1,1165 @@
...
+            messages.add_message(request, messages.INFO,r.json()["message"], "danger")
+
+            return redirect("App:enroll")
+
+    return render(request,"enroll.html")
+
+```
+In the code above, we collected the typing pattern enrolled by the user and stored in the `tp` variable. We then requested the `auto` endpoint of the TypingDNA API to verify if the typing pattern provided matches the typing patterns enrolled for that user . We then checked if the status code of the request was successful (status code 200). If this was the case, we would update the user data in our database to indicate they have been enrolled. Then we will mark the current logged-in session as authenticated with `session["typingdna_auth"] = True` then redirect the user to their dashboard. However, if the authentication was unsuccessful (due to an error from TypingDNA) we will prompt the error message to the user and let them retry the enrollment process.
+
+After successfully enrolling to TypingDNA, we should be redirected to a page similar to the one in the image below:
```
Comment (paulodhiambo) [inline articles/biometric-2fa-in-a-django-application-with-typingdna/index.md:631]: Change "After successfully enrolling to TypingDNA, ..." to "After successfully enrolling in TypingDNA, ..."

File: articles/biometric-2fa-in-a-django-application-with-typingdna/index.md [hunk 4]
```diff
@@ -0,0 +1,1165 @@
...
+
+After registering and enrolling our typing patterns with TypingDNA, we can now login and implement our two-factor authentication.
+
+![login](/engineering-education/biometric-2fa-in-a-django-application-with-typingdna/login.png)
+
+In the image above, our `login` view checks if the user is set in session. If the user is set then the user is redirected to view all videos page. However, if the user is not set in the session, then the user is redirected to verify his/her 2FA with their typing pattern.
+
+## Building The Verify Page
+
+We want to build our verification HTML page where we will verify user identities using TypingDNA biometrics. We will also create a fall-back verification page where users can verify. Lets update our `verify.html` file as seen in the code below:
```
Comment (paulodhiambo) [inline articles/biometric-2fa-in-a-django-application-with-typingdna/index.md:647]: Change "Lets update our..." to "Let's update our..."

**Review by paulodhiambo — CHANGES_REQUESTED** [review]
@Nomzy-kush Good article though you need to pass the article through a grammar checker (Grammarly) since it has several grammar mistakes that can be corrected by grammar checker. Ping me once you have corrected the grammar mistakes.

**Comment by Nomzy-kush** [comment]
@paulodhiambo I have made the necessary changes you asked me to

**Comment by hectorkambow** [comment]
plagiarism check done

File: articles/biometric-2fa-in-a-django-application-with-typingdna/index.md [hunk 5]
```diff
@@ -0,0 +1,1158 @@
...
+
+Have you ever wanted to implement biometric two-factor authentication in your Django web application? If your answer is yes, then this article is just what you need to get started.
+
+TypingDNA is a two-factor authentication API that easily integrates biometrics authentication to secure user accounts on your website using powerful typing analysis. TypingDNA typically helps you learn a user’s typing pattern, match them, and then use it to authenticate users after login. This double user verification is what is known as a “two-factor authentication”.
+
+This article shows a step-by-step guide to integrating TypingDNA services into a Django web application. To learn more, visit their official website [TypingDNA](https://www.typingdna.com)
+
+### Signing Up with TypingDNA
+
+To get started, we need to create an account on their website at [TypingDNA signup](https://www.typingdna.com/clients/signup). After creating and setting up our account, we should see a page like the one in the image below. Copy your `api_key` and `secret_key` and store them somewhere they are safe and easily retrievable.
```
Comment (paulodhiambo) [inline articles/biometric-2fa-in-a-django-application-with-typingdna/index.md:11]: Please add the image you're talking about here.

File: articles/biometric-2fa-in-a-django-application-with-typingdna/index.md [hunk 6]
```diff
@@ -0,0 +1,1158 @@
...
+
+        })
+
+
+```
+
+
+
+
+Next, we need to create a `UserProfile` table for use our users to start verifying typing patterns captured typing patterns containing extra details for TypingDNA authentication users in our database. Update the `models.py` file with the code below:
```
Comment (paulodhiambo) [inline articles/biometric-2fa-in-a-django-application-with-typingdna/index.md:122]: This sentence is not clear.

**Review by paulodhiambo — CHANGES_REQUESTED** [review]
@Nomzy-kush  make the requested changes then ping me when done.

**Comment by Nomzy-kush** [comment]
@paulodhiambo I have made all the requested changes

**Comment by paulodhiambo** [comment]
@Nomzy-kush amazing article. Take note of the changes I have made. @hectorkambow you can give this article a final review.

**Comment by Nomzy-kush** [comment]
This fixes #1727

**Pull request merged**